File: Jump Game II.java
class Solution {
    public int jump(int[] nums) {
        int[] dp=new int[nums.length];
        Arrays.fill(dp,-1);
        int val=helper(nums,0,dp);
        return val;
    }
    public int helper(int[] nums,int i,int[] dp){
        if(i==nums.length-1){
            return 0;
        }
        if(dp[i]!=-1){
            return dp[i];
        }
        int ans=100000;
        for(int j=1;j<=nums[i];j++){
            if(i+j<nums.length){
                ans=Integer.min(ans,helper(nums,i+j,dp)+1);
            }
        }
        dp[i]=ans;
        return ans;
        
    }
}

File: Jump Game II.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        int curr=0;
        int farthest=0;
        int jump=0;

        int n = nums.size();
        for(int i=0;i<n-1;i++)
         {
             farthest = max(farthest, nums[i]+i);

             if(i==curr)
              {
                  jump++;
                  curr= farthest;
              }
         }
       return jump;
    }
};
